src/app/clusters/time-synchronization-server/time-synchronization-server.cpp

```diff
@@ -0,0 +1,980 @@
+CHIP_ERROR TimeSynchronizationServer::ClearTimeZone()
+{
+    mTimeZoneObj.size         = 1; // one default time zone item is needed
+    mTimeZoneObj.timeZoneList = Span<TimeZoneStore>(mTz);
+    for (auto & tzObj : mTimeZoneObj.timeZoneList)
+    {
+        memset(tzObj.name, 0, sizeof(tzObj.name));
+        tzObj.timeZone = { .offset = 0, .validAt = 0, .name = MakeOptional(CharSpan(tzObj.name, sizeof(tzObj.name))) };
+    }
+    return mTimeSyncDataProvider.StoreTimeZone(GetTimeZone());
+}
+
+CHIP_ERROR TimeSynchronizationServer::ClearDSTOffset()
+{
+    mDstOffsetObj.size = 0;
+    return mTimeSyncDataProvider.ClearDSTOffset();
+}
+
+DataModel::Nullable<Structs::TrustedTimeSourceStruct::Type> & TimeSynchronizationServer::GetTrustedTimeSource()
+{
+    return mTrustedTimeSource;
+}
+
+CHIP_ERROR TimeSynchronizationServer::GetDefaultNtp(MutableCharSpan & dntp)
+{
+    return mTimeSyncDataProvider.LoadDefaultNtp(dntp);
+}
+
+Span<TimeZoneStore> TimeSynchronizationServer::GetTimeZone()
+{
+    return mTimeZoneObj.timeZoneList.SubSpan(0, mTimeZoneObj.size);
+}
+
+DataModel::List<Structs::DSTOffsetStruct::Type> TimeSynchronizationServer::GetDSTOffset()
+{
+    return mDstOffsetObj.dstOffsetList.SubSpan(0, mDstOffsetObj.size);
+}
+
+void TimeSynchronizationServer::ScheduleDelayedAction(System::Clock::Seconds32 delay, System::TimerCompleteCallback action,
+                                                      void * aAppState)
+{
+    if (CHIP_NO_ERROR != SystemLayer().StartTimer(std::chrono::duration_cast<System::Clock::Timeout>(delay), action, aAppState))
+    {
+        ChipLogError(Zcl, "Time Synchronization failed to schedule timer.");
+    }
+}
+
+CHIP_ERROR TimeSynchronizationServer::SetUTCTime(EndpointId ep, uint64_t utcTime, GranularityEnum granularity,
+                                                 TimeSourceEnum source)
+{
+    ReturnErrorOnFailure(UpdateUTCTime(utcTime));
+    mGranularity = granularity;
+    if (EMBER_ZCL_STATUS_SUCCESS != TimeSource::Set(ep, source))
+    {
+        ChipLogError(Zcl, "Writing TimeSource failed.");
+        return CHIP_IM_GLOBAL_STATUS(Failure);
+    }
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR TimeSynchronizationServer::GetLocalTime(EndpointId ep, DataModel::Nullable<uint64_t> & localTime)
+{
+    int64_t timeZoneOffset = 0, dstOffset = 0;
+    System::Clock::Microseconds64 utcTime;
+    uint64_t chipEpochTime;
+    VerifyOrReturnError(TimeState::kInvalid != GetUpdatedDSTOffsetState(), CHIP_ERROR_INVALID_TIME);
+    ReturnErrorOnFailure(System::SystemClock().GetClock_RealTime(utcTime));
+    VerifyOrReturnError(UnixEpochToChipEpochMicro(utcTime.count(), chipEpochTime), CHIP_ERROR_INVALID_TIME);
+    if (TimeState::kChanged == GetUpdatedTimeZoneState())
+    {
+        sendTimeZoneStatusEvent(ep);
+    }
+    VerifyOrReturnError(GetTimeZone().size() != 0, CHIP_ERROR_INVALID_TIME);
+    const auto & tzStore = GetTimeZone()[0];
+    timeZoneOffset       = static_cast<int64_t>(tzStore.timeZone.offset);
```

Comment:
Why is this cast needed?  Offset should already be signed, which means this assignment should succeed without casts.  If it does not, we _want_ the compiler to yell at us, because something has gone very wrong!

Similar for `dst.offset` below.

Followup is OK to fix this.

Reply:
Tracked in https://github.com/project-chip/connectedhomeip/issues/27575, resolving this comment for now, further followup on the issue.